src/udev/iocost/iocost.c

```diff
@@ -0,0 +1,328 @@
+
+        r = sd_device_get_property_value(device, model_key, &model);
+        if (r == -ENOENT) {
+                log_debug_errno(r, "No value found for key %s, skipping iocost logic.", model_key);
+                return 0;
+        }
+        if (r < 0)
+                return log_error_errno(r, "Failed to obtain model for iocost solution from device: %m");
+
+        *ret_qos = qos;
+        *ret_model = model;
+
+        return 0;
+}
+
+static int apply_solution_for_path(const char *path, const char *name) {
+        _cleanup_(sd_device_unrefp) sd_device *device = NULL;
+        _cleanup_free_ char *qos = NULL, *model = NULL;
+        const char *qos_params = NULL, *model_params = NULL;
+        dev_t devnum;
+        int r;
+
+        r = sd_device_new_from_path(&device, path);
+        if (r < 0)
+                return log_error_errno(r, "Error looking up device: %m");
+
+        r = query_named_solution(device, name, &model_params, &qos_params);
+        if (r < 0)
+                return r;
+
+        r = sd_device_get_devnum(device, &devnum);
+        if (r < 0)
+                return log_error_errno(r, "Error getting devnum for device %s: %m", path);
+
+        if (asprintf(&qos, DEVNUM_FORMAT_STR " enable=1 ctrl=user %s", DEVNUM_FORMAT_VAL(devnum), qos_params) < 0)
+                return log_oom();
+
+        if (asprintf(&model, DEVNUM_FORMAT_STR " model=linear ctrl=user %s", DEVNUM_FORMAT_VAL(devnum), model_params) < 0)
+                return log_oom();
+
+        log_debug("Applying iocost parameters to %s using solution '%s'\n"
+                        "\tio.cost.qos: %s\n"
+                        "\tio.cost.model: %s\n", path, name ?: "default", qos, model);
+
+        r = cg_set_attribute("io", NULL, "io.cost.qos", qos);
+        if (r < 0) {
+                log_full_errno(r == -ENOENT ? LOG_DEBUG : LOG_ERR, r, "Failed to set io.cost.qos: %m");
+                return r == -ENOENT ? 0 : r;
+        }
+
+        r = cg_set_attribute("io", NULL, "io.cost.model", model);
+        if (r < 0) {
+                log_full_errno(r == -ENOENT ? LOG_DEBUG : LOG_ERR, r, "Failed to set io.cost.model: %m");
+                return r == -ENOENT ? 0 : r;
+        }
+
+        return 0;
+}
+
+static int query_solutions_for_path(const char *path) {
+        _cleanup_(sd_device_unrefp) sd_device *device = NULL;
+        _cleanup_strv_free_ char **solutions = NULL;
+        const char *default_solution = NULL;
+        const char *model_name = NULL;
+        int r;
+
+        r = sd_device_new_from_path(&device, path);
+        if (r < 0)
+                return log_error_errno(r, "Error looking up device: %m");
+
+        if ((sd_device_get_property_value(device, "ID_MODEL_FROM_DATABASE", &model_name) < 0 &&
+             sd_device_get_property_value(device, "ID_MODEL", &model_name) < 0))
+                return log_error_errno(SYNTHETIC_ERRNO(ENOENT), "Model name for device %s is unknown", path);
```

Comment:
would prefer if we'd pass the actual return values of sd_device_get_property_value() as first arg, instead of making up ENOENT.

also log_device_error_errno

Reply:
Done in the latest push.